### PR TITLE
crimson/os/seastore: add set/get attrs and write/read meta methods for seastore

### PR DESCRIPTION
**Pull request by xxhdx1985126**
Signed-off-by: Xuehan Xu <xxhdx1985126@gmail.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/test/crimson/seastore/test_seastore.cc [hunk 1]
```diff
@@ -84,6 +95,22 @@ struct seastore_test_t :
...
+      SeaStore &seastore,
+      bufferlist& val) {
+      CTransaction t;
+      t.setattr(cid, oid, OI_ATTR, val);
+      seastore.do_transaction(
+        coll,
+        std::move(t)).get0();
+    }
+
+    SeaStore::attrs_t get_attr_oi(
```
Comment (ronen-fr) [inline src/test/crimson/seastore/test_seastore.cc:108]: doesn't that make the function unusable in "regular" (not a separate thread) seastar context?

File: src/test/crimson/seastore/test_seastore.cc [hunk 2]
```diff
@@ -185,6 +196,20 @@ TEST_F(seastore_test_t, collection_create_list_remove)
   });
 }
 
+TEST_F(seastore_test_t, meta) {
+  run_async([this] {
+    set_meta("key1", "value1");
+    set_meta("key2", "value2");
+
+    auto ret1 = get_meta("key1");
```
Comment (tchaikov) [inline src/test/crimson/seastore/test_seastore.cc:220]: nit, might want to put
```c++
const auto [ret1, value1] = get_meta("key1");
```

File: src/test/crimson/seastore/test_seastore.cc [hunk 3]
```diff
@@ -185,6 +196,20 @@ TEST_F(seastore_test_t, collection_create_list_remove)
...
+    auto ret2 = get_meta("key2");
+    EXPECT_EQ(std::get<0>(ret1), 0);
+    EXPECT_EQ(std::get<0>(ret2), 0);
+    EXPECT_TRUE(std::get<1>(ret1) == "value1");
```
Comment (tchaikov) [inline src/test/crimson/seastore/test_seastore.cc:224]: nit, 
```c++
EXPECT_EQ(value1 == "value1");
```
with the proposed change.

File: src/crimson/os/seastore/cache.cc [hunk 4]
```diff
@@ -193,8 +193,12 @@ std::optional<record_t> Cache::try_construct_record(Transaction &t)
 {
   // First, validate read set
   for (auto &i: t.read_set) {
-    if (i->state == CachedExtent::extent_state_t::INVALID)
+    if (i->state == CachedExtent::extent_state_t::INVALID) {
+      logger().debug("{}, invalid cachedextent in read_set, {}",
+		     __func__,
+		     *i);
```
Comment (tchaikov) [inline src/crimson/os/seastore/cache.cc:199]: i'd suggest move this change into its own commit. but i am inclined to drop it. as it happens all the time in our seastore implementation.

@athanatos what do you think?

File: src/crimson/os/seastore/root_block.h [hunk 5]
```diff
@@ -66,6 +68,11 @@ struct RootBlock : CachedExtent {
     ceph::bufferlist bl = _bl;
     bl.rebuild();
     root = *reinterpret_cast<const root_t*>(bl.front().c_str());
+    if (root.have_meta) {
+      ceph::bufferlist bl;
```
Comment (tchaikov) [inline src/crimson/os/seastore/root_block.h:72]: would be better if we don't shadow the variable in the parent scope.

File: src/crimson/os/seastore/root_block.h [hunk 6]
```diff
@@ -38,6 +38,8 @@ struct RootBlock : CachedExtent {
 
   root_t root;
 
+  std::map<std::string, std::string> meta;
```
Comment (tchaikov) [inline src/crimson/os/seastore/root_block.h:41]: might want to use `std::map<string_view, string_view> meta`? as the content of `meta` *can* be backed by `root_t::meta[]`.

File: src/crimson/os/seastore/root_block.h [hunk 7]
```diff
@@ -66,6 +68,11 @@ struct RootBlock : CachedExtent {
     ceph::bufferlist bl = _bl;
     bl.rebuild();
     root = *reinterpret_cast<const root_t*>(bl.front().c_str());
+    if (root.have_meta) {
```
Comment (tchaikov) [inline src/crimson/os/seastore/root_block.h:71]: i am not sure i am missing something obvious. but what if a delta is intended to add meta to a root without meta?

File: src/crimson/os/seastore/transaction_manager.h [hunk 8]
```diff
@@ -322,6 +322,28 @@ class TransactionManager : public SegmentCleaner::ExtentCallbackInterface {
...
+  }
+
+  void update_root_meta(Transaction& t,
+                        const std::string& key,
+                        const std::string& value) {
+    auto root = cache->get_root_fast(t);
+    root = cache->duplicate_for_write(t, root)->cast<RootBlock>();
+    root->meta[key] = value;
+    ceph::bufferlist bl(1);
+    bl.rebuild(ceph::buffer::ptr_node::create(&root->get_root().meta[0], root_t::MAX_META_LENGTH));
```
Comment (tchaikov) [inline src/crimson/os/seastore/transaction_manager.h:343]: might want to error out if the total size of meta map exceeds the capacity of `meta[]`? also, i think we can save the space using `string_view` here.

File: src/crimson/os/seastore/cache.cc (hunk 4 above)
Comment (athanatos) [reply src/crimson/os/seastore/cache.cc:199]: +1 to moving into a separate commit.  I'm fine with adding the debugging if it was useful.

File: src/crimson/os/seastore/root_block.h (hunk 6 above)
Comment (athanatos) [reply src/crimson/os/seastore/root_block.h:41]: No, keep it simple.

File: src/crimson/os/seastore/seastore.cc [hunk 9]
```diff
@@ -1,5 +1,5 @@
-// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
-// vim: ts=8 sw=2 smarttab
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:nil -*-
+// vim: ts=8 sw=2 smarttab expandtab
```
Comment (athanatos) [inline src/crimson/os/seastore/seastore.cc:2]: You've changed this rule, but not actually updated the whitespace elsewhere in this file.  If you really want to do this, please do it in its own PR and update all of the files in seastore/ at the same time.

File: src/crimson/os/seastore/seastore.cc [hunk 10]
```diff
@@ -199,7 +199,16 @@ SeaStore::get_attrs_ertr::future<SeaStore::attrs_t> SeaStore::get_attrs(
   auto c = static_cast<SeastoreCollection*>(ch.get());
   logger().debug("{} {} {}",
 		 __func__, c->get_cid(), oid);
-  return crimson::ct_error::enoent::make();
+  return repeat_with_onode<SeaStore::attrs_t>(
```
Comment (athanatos) [inline src/crimson/os/seastore/seastore.cc:202]: Don't need the SeaStore:: here.

File: src/crimson/os/seastore/transaction_manager.h [hunk 11]
```diff
@@ -1,5 +1,5 @@
-// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
-// vim: ts=8 sw=2 smarttab
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:nil -*-
+// vim: ts=8 sw=2 smarttab expandtab
```
Comment (athanatos) [inline src/crimson/os/seastore/transaction_manager.h:2]: Again, please remove all of these until you plan on updating the tabbing.

File: src/crimson/os/seastore/root_block.h (hunk 6 above)
Comment (athanatos) [reply src/crimson/os/seastore/root_block.h:41]: Actually, we don't want to be decoding/encoding this every time we write out the root node.  Can we just keep this in memory as a char[] as with root_t?

File: src/crimson/os/seastore/transaction_manager.h (hunk 8 above)
Comment (athanatos) [inline src/crimson/os/seastore/transaction_manager.h:343]: This line is more than 80 characters long.

**Comment by athanatos** [comment]
We still need SS_ATTR, right?

File: src/crimson/os/seastore/onode.h [hunk 12]
```diff
@@ -14,8 +14,14 @@
 namespace crimson::os::seastore {
 
 struct onode_layout_t {
+  // around 350 bytes for fixed fields in object_info_t,
+  // the left are for the variable-sized fields like oid
+  static constexpr int MAX_OI_LENGTH = 1024;
```
Comment (cyx1231st) [inline src/crimson/os/seastore/onode.h:19]: I think there are two considerations about the size of `onode_layout_t`:
1.  The size limit of the key(`ghobject_t`) and value(`onode_layout_t`) to not break the onode-tree node-level split/merge operations;
2. The current implementation for `onode_layout_t` doesn't prefer large size, because inplace update would require to imbed the entire `sizeof(onode_layout_t)` into the seastore delta;

For 1. according to https://github.com/ceph/ceph/blob/master/src/crimson/os/seastore/onode_manager/staged-fltree/node_layout.h#L414-L447. If correct, we would need to set a proper size upper-bounds so that:
4 * max-key-size + 2 * max-value-size < 4KiB (node size),  for example both set to 640B.

For 2. I think currently we might need to keep the size of `onode_layout_t` as small as possible.

File: src/crimson/os/seastore/onode.h [hunk 13]
```diff
@@ -14,8 +14,14 @@
 namespace crimson::os::seastore {
 
 struct onode_layout_t {
+  // around 350 bytes for fixed fields in object_info_t,
+  // the left are for the variable-sized fields like oid
```
Comment (cyx1231st) [inline src/crimson/os/seastore/onode.h:18]: I don't think we need extra space in `onode_layout_t` to store oid, as it is already present in the tree as the key. For example, we can get the oid from Cursor::get_ghobj(), see https://github.com/ceph/ceph/blob/master/src/crimson/os/seastore/onode_manager/staged-fltree/tree.h#L80-L84

File: src/crimson/os/seastore/onode.h [hunk 14]
```diff
@@ -14,8 +14,14 @@
...
+  static constexpr int MAX_OI_LENGTH = 1024;
+
   ceph_le32 size{0};
```
Comment (cyx1231st) [inline src/crimson/os/seastore/onode.h:21]: I don't think we need `onode_layout_t::size`, as it is already implemented in `Value::get_payload_size()`, see https://github.com/ceph/ceph/blob/master/src/crimson/os/seastore/onode_manager/staged-fltree/value.h#L180-L183

File: src/crimson/os/seastore/seastore.cc [hunk 15]
```diff
@@ -760,6 +769,18 @@ SeaStore::tm_ret SeaStore::_setattrs(
...
+  auto pattr = aset.find(OI_ATTR);
+  if (pattr != aset.end()) {
+    if (__builtin_expect(
+        pattr->second.length() > onode_layout_t::MAX_OI_LENGTH,
+        false)) {
+      logger().error("{} onode={} oi too long!");
+      return crimson::ct_error::input_output_error::make();
+    }
+    layout.oi_size = pattr->second.length();
+    pattr->second.copy_out(0, pattr->second.length(), &layout.oi[0]);
```
Comment (cyx1231st) [inline src/crimson/os/seastore/seastore.cc:782]: Worth noting here that the actual size of `onode_layout_t` will not fit to the dynamic OI size, which will always be `sizeof(onode_layout_t)`, see https://github.com/ceph/ceph/blob/master/src/crimson/os/seastore/onode_manager/staged-fltree/fltree_onode_manager.cc#L46.

We can only adjust the size after implement `Value::trim()` and `Value::extend()`, see https://github.com/ceph/ceph/blob/master/src/crimson/os/seastore/onode_manager/staged-fltree/value.h#L191-L195.

File: src/crimson/os/seastore/onode.h (hunk 13 above)
Comment (athanatos) [reply src/crimson/os/seastore/onode.h:18]: That's out of scope here.  The issue is that the OSD's own object_info_t structure (which it stores in the OI_ATTR) encodes the full hobject_t.  A good next project would be to rework the OSD's usage of this interface to remove that requirement and thus shrink the object_info_t.

Comment (cyx1231st) [reply src/crimson/os/seastore/onode.h:18]: Hmm, looks `object_info_t` is encoded outside `ObjectStore`, need to adjust OSD logic to exclude oid from the OI buffer.

File: src/crimson/os/seastore/root_block.h (hunk 7 above)
Comment (xxhdx1985126) [reply src/crimson/os/seastore/root_block.h:71]: I thought that the whole root_t would be inclosed in the delta as indicated by the code in RootBlock::get_delta(), am I right?

File: src/crimson/os/seastore/onode.h (hunk 13 above)
Comment (athanatos) [reply src/crimson/os/seastore/onode.h:18]: Right.

File: src/crimson/os/seastore/seastore.cc (hunk 15 above)
Comment (athanatos) [reply src/crimson/os/seastore/seastore.cc:782]: A fixed size will suffice for now.  Can we use large leaf nodes for now (64?)?

Comment (xxhdx1985126) [reply src/crimson/os/seastore/seastore.cc:782]: Maybe I can store OI_ATTR and SS_ATTR in other extent(s) and add a pointer in onode_layout_t, so that the onode tree won't need to be adjusted.

Comment (athanatos) [reply src/crimson/os/seastore/seastore.cc:782]: Hmm.  The thing is that every single osd object access hits these bits of metadata.  I don't think it's actually more efficient to move them to a secondary location.  However, whatever is most expedient will do for now -- we can improve it later.

Comment (cyx1231st) [reply src/crimson/os/seastore/seastore.cc:782]: I think we can, but we might need to reduce the leaf node size in the future after oid is excluded from value. Larger node means more memcpy and disk I/O for most of tree operations.

----

IIUC TransactionManager already supports 64K extent managements, so the major work would be in the tree implementation.

I think at least we need to remove the hard-coded node size [1], set it to 64K somewhere else and see whether it works immediately. I don't see any size-specific logic that won't work after the change, but very likely there are unexpected corners.

Leaf node split tests [2] can be bothering, because each test layout is manually constructed based on the 4K leaf node to validate 45 different split scenarios. I'm hesitate to rewrite them under 64K, maybe we can just leave these tests with 4K leaf nodes.

There're a number of internal changes pending in the erase/merge feature [3]. It would be better to implement 64K leaf after that. But I think I can reprioritize if needed.

[1] https://github.com/ceph/ceph/blob/master/src/crimson/os/seastore/onode_manager/staged-fltree/stages/node_stage_layout.h#L133
[2] https://github.com/ceph/ceph/blob/master/src/test/crimson/seastore/onode_tree/test_staged_fltree.cc#L513
[3] https://github.com/cyx1231st/ceph/commits/wip-seastore-onode-tree-delete

**Comment by xxhdx1985126** [comment]
@tchaikov @athanatos @cyx1231st I've corrected the code as suggested, please take a look:-)

File: src/crimson/os/seastore/onode_manager/staged-fltree/fltree_onode_manager.cc [hunk 16]
```diff
@@ -41,6 +42,10 @@ FLTreeOnodeManager::get_or_create_onode_ret
 FLTreeOnodeManager::get_or_create_onode(
   Transaction &trans,
   const ghobject_t &hoid) {
+  if (hoid.hobj.oid.name.length() + hoid.hobj.nspace.length()
+      > key_view_t::MAX_NS_OID_LENGTH) {
```
Comment (athanatos) [inline src/crimson/os/seastore/onode_manager/staged-fltree/fltree_onode_manager.cc:46]: @cyx1231st This is something we plan on fixing, right?  There are osd level limits on namespace and name length, but other than that this implementation must deal with any length.

File: src/crimson/os/seastore/seastore.cc [hunk 17]
```diff
@@ -760,6 +769,30 @@ SeaStore::tm_ret SeaStore::_setattrs(
...
+  if (ss_attr != aset.end()) {
+    if (__builtin_expect(
+        ss_attr->second.length() > onode_layout_t::MAX_SS_LENGTH,
+        false)) {
+      logger().error("{} onode={} ss attr too long!");
+      return crimson::ct_error::input_output_error::make();
+    }
+    layout.oi_size = ss_attr->second.length();
+    ss_attr->second.copy_out(0, ss_attr->second.length(), &layout.oi[0]);
+  }
```
Comment (athanatos) [inline src/crimson/os/seastore/seastore.cc:795]: Return an error if there are any other xattrs specified -- other xattrs are *legal*, we just aren't dealing with them in this PR.

Comment (athanatos) [reply src/crimson/os/seastore/seastore.cc:795]: Or actually, just assert since the next step will be to handle that case.

File: src/crimson/os/seastore/onode.h (hunk 14 above)
Comment (athanatos) [reply src/crimson/os/seastore/onode.h:29]: That size isn't the onode size, it's the size of the object itself.

Comment (athanatos) [reply src/crimson/os/seastore/onode.h:29]: ...or will be shortly.

File: src/crimson/os/seastore/onode_manager/staged-fltree/fltree_onode_manager.cc (hunk 16 above)
Comment (cyx1231st) [reply src/crimson/os/seastore/onode_manager/staged-fltree/fltree_onode_manager.cc:46]: Yes, as long as we can figure out the upper-bound sizes of keys and values, we can then adjust the sizes of internal nodes and leaf nodes statically.

Dealing with **any length** of value without upper-bound is tricky I think. At least it seems unintuitive to me if node sizes need to be dynamic and are inhomogeneous at the same level.

Comment (athanatos) [reply src/crimson/os/seastore/onode_manager/staged-fltree/fltree_onode_manager.cc:46]: osd_max_object_name_len (2k by default) and object_max_object_namespace_len (256) are the configs.  We must handle any hobject within those limits.

Comment (cyx1231st) [reply src/crimson/os/seastore/onode_manager/staged-fltree/fltree_onode_manager.cc:46]: These means the key has upperbound, so they should be fine.

How about the imbedded value (`onode_layout_t`)? Does it also has upperbound?

Comment (athanatos) [reply src/crimson/os/seastore/onode_manager/staged-fltree/fltree_onode_manager.cc:46]: Yes, though we'll want it to be large enough to contain OI_ATTR and SS_ATTR for normal rbd blocks (with an xattr btree for other attrs or if those are too large).

Comment (cyx1231st) [reply src/crimson/os/seastore/onode_manager/staged-fltree/fltree_onode_manager.cc:46]: OK, this should be fine I believe.

File: src/crimson/os/seastore/onode_manager/staged-fltree/stages/key_layout.h [hunk 18]
```diff
@@ -555,6 +556,11 @@ inline std::ostream& operator<<(std::ostream& os, const key_hobj_t& key) {
  */
 class key_view_t {
  public:
+  static constexpr int MAX_NS_OID_LENGTH =
+    (4096 - sizeof(onode_layout_t) * 2) / 4
+    - sizeof(shard_pool_t) - sizeof(crush_t) - sizeof(snap_gen_t)
+    - 8; // size of length field of oid and ns
```
Comment (cyx1231st) [inline src/crimson/os/seastore/onode_manager/staged-fltree/stages/key_layout.h:562]: To be precise this should include the overhead introduced by internal indexing and node/stage/value headers.

But I'm fine with this because actually `MAX_NS_OID_LENGTH` is defined by OSD and we will need to adjust the node size accordingly.

Maybe it is appropriate to add a `// FIXME` here?

File: src/crimson/os/seastore/onode.h (hunk 13 above)
Comment (cyx1231st) [reply src/crimson/os/seastore/onode.h:18]: @xxhdx1985126 could add a FIXME here to remember that we want to exclude the oid in the future?

Comment (xxhdx1985126) [reply src/crimson/os/seastore/onode.h:18]: Ok

File: src/crimson/os/seastore/onode_manager/staged-fltree/stages/key_layout.h (hunk 18 above)
Comment (xxhdx1985126) [reply src/crimson/os/seastore/onode_manager/staged-fltree/stages/key_layout.h:562]: @cyx1231st Do you mean that we drop this change and add a FIXME for a change?

Comment (cyx1231st) [reply src/crimson/os/seastore/onode_manager/staged-fltree/stages/key_layout.h:562]: No, add a FIXME to remember that `MAX_NS_OID_LENGTH` should be defined by OSD in the future.

Comment (xxhdx1985126) [reply src/crimson/os/seastore/onode_manager/staged-fltree/stages/key_layout.h:562]: ok

**Comment by tchaikov** [comment]
jenkins test api

File: src/crimson/os/seastore/transaction_manager.h [hunk 19]
```diff
@@ -322,6 +322,52 @@ class TransactionManager : public SegmentCleaner::ExtentCallbackInterface {
...
+    auto root = cache->get_root_fast(t);
+    root = cache->duplicate_for_write(t, root)->cast<RootBlock>();
+    root->meta[key] = value;
+
+    // calculate meta size
+    // TODO: we probably need a uniformal interface for calcuting
+    // the encoded size of data structures
+    uint32_t meta_size = 4; // initial 4 bytes for std::map size
+    for (auto& [key, val] : root->meta) {
+      meta_size += 8 + key.length() + val.length();
```
[the remark this answers is not in the document]
Comment (xxhdx1985126) [reply src/crimson/os/seastore/transaction_manager.h:356]: size of fields for length of the key and val

[the remark this answers is not in the document]
Comment (liu-chunmei) [reply src/crimson/os/seastore/transaction_manager.h:408]: could you please add a comment here?

[the remark this answers is not in the document]
Comment (xxhdx1985126) [reply src/crimson/os/seastore/transaction_manager.h:408]: ok, that'll be done

**Comment by xxhdx1985126** [comment]
@tchaikov @athanatos @cyx1231st @liu-chunmei I think this pr has been modified according to your suggestions, please take a look, again, thanks:-)

**Review by liu-chunmei — APPROVED** [review]

**Comment by athanatos** [comment]
@xxhdx1985126 make check seems to be failing, probably because of another PR merge.

**Review by athanatos — APPROVED** [review]

**Comment by athanatos** [comment]
LGTM once build fixed.

**Review by tchaikov — APPROVED** [review]
